File: app/money.py
"""Arithmetique monetaire. Tout montant du projet passe par ici.

EX-07 : aucun calcul n'est confie au modele. Decimal + quantize a 2 decimales,
arrondi bancaire ROUND_HALF_UP comme en comptabilite marocaine.
"""
from decimal import Decimal, ROUND_HALF_UP, InvalidOperation
import re

CENT = Decimal("0.01")
TOLERANCE = Decimal("0.02")  # tolerance d'arrondi sur un controle de coherence


def d(value) -> Decimal:
    """Convertit en Decimal. Renvoie None si inexploitable."""
    if value is None:
        return None
    if isinstance(value, Decimal):
        return value
    try:
        return Decimal(str(value))
    except (InvalidOperation, ValueError):
        return None


def q(value) -> Decimal:
    """Arrondit a 2 decimales."""
    value = d(value)
    return None if value is None else value.quantize(CENT, rounding=ROUND_HALF_UP)


def parse_montant(texte: str) -> Decimal:
    """Lit un montant dans du texte brut ou de l'OCR.

    Gere les espaces insecables, la virgule decimale, le signe negatif des
    avoirs, et les caracteres parasites que l'OCR colle aux chiffres
    ('34 828.0!' -> 34828.0).
    """
    if texte is None:
        return None
    t = str(texte).replace("\u00a0", " ").replace("\u202f", " ").strip()
    negatif = t.startswith("-") or t.startswith("(")
    t = re.sub(r"[^0-9.,]", "", t)
    if not t:
        return None
    # separateur decimal = dernier . ou , suivi de 1 ou 2 chiffres
    m = re.search(r"[.,](\d{1,2})$", t)
    if m:
        entier = re.sub(r"[^0-9]", "", t[: m.start()])
        val = f"{entier or '0'}.{m.group(1)}"
    else:
        val = re.sub(r"[^0-9]", "", t)
    if not val or val == ".":
        return None
    montant = d(val)
    if montant is None:
        return None
    return -montant if negatif else montant


def tva_attendue(ht: Decimal, taux: int) -> Decimal:
    """TVA theorique pour une base HT et un taux entier."""
    return q(d(ht) * d(taux) / Decimal(100))


def coherent(ht: Decimal, tva: Decimal, ttc: Decimal) -> bool:
    """HT + TVA == TTC, a la tolerance d'arrondi pres."""
    if None in (ht, tva, ttc):
        return False
    return abs((q(ht) + q(tva)) - q(ttc)) <= TOLERANCE

File: api/src/money.ts
/**
 * Arithmetique monetaire. Tout montant du projet passe par ici.
 *
 * Le cahier des charges impose decimal.js et un stockage en `numeric` cote
 * Postgres : aucun montant ne transite par un flottant, et aucun n'est calcule
 * par le modele.
 */
import { Decimal } from "decimal.js";

// 2 decimales, arrondi comptable (0,005 monte).
Decimal.set({ precision: 28, rounding: Decimal.ROUND_HALF_UP });

export const ZERO = new Decimal(0);
export const TOLERANCE = new Decimal("0.02");

export type Montant = Decimal;

/** Arrondit a deux decimales. */
export function q(valeur: Decimal.Value): Montant {
  return new Decimal(valeur).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
}

/**
 * Lit un montant dans du texte brut ou de l'OCR.
 * Gere les espaces insecables, la virgule decimale, le signe negatif des
 * avoirs, et les caracteres parasites colles aux chiffres ("34 828.0!").
 */
export function lireMontant(texte: string | null | undefined): Montant | null {
  if (texte === null || texte === undefined) return null;
  let t = String(texte).replace(/\u00a0|\u202f/g, " ").trim();
  const negatif = t.startsWith("-") || t.startsWith("(");
  t = t.replace(/[^0-9.,]/g, "");
  if (!t) return null;

  // Separateur decimal : dernier . ou , suivi d'un ou deux chiffres.
  const decimale = t.match(/[.,](\d{1,2})$/);
  let valeur: string;
  if (decimale) {
    const entier = t.slice(0, decimale.index).replace(/[^0-9]/g, "");
    valeur = `${entier || "0"}.${decimale[1]}`;
  } else {
    valeur = t.replace(/[^0-9]/g, "");
  }
  if (!valeur) return null;

  try {
    const montant = new Decimal(valeur);
    return negatif ? montant.neg() : montant;
  } catch {
    return null;
  }
}

/** TVA theorique pour une base HT et un taux entier. */
export function tvaAttendue(ht: Decimal.Value, taux: number): Montant {
  return q(new Decimal(ht).times(taux).dividedBy(100));
}

/** HT + TVA vaut-il TTC, a la tolerance d'arrondi pres ? */
export function coherent(
  ht: Montant | null,
  tva: Montant | null,
  ttc: Montant | null
): boolean {
  if (!ht || !tva || !ttc) return false;
  return q(ht).plus(q(tva)).minus(q(ttc)).abs().lessThanOrEqualTo(TOLERANCE);
}

/** Forme d'echange et de stockage : une chaine, jamais un flottant. */
export const texte = (montant: Montant | null): string | null =>
  montant === null ? null : q(montant).toFixed(2);
